#                                                                                                                                                         Awe Sparks Prosociality in Children  
#                                                                                                                                         Instructor: Chen Wanting、Xu Zhaosuyang、Zhou Yujie  
Stamkou, E., Brummelman, E., Dunham, R. L., Nikolić, M., & Keltner, D. (2023). Awe sparks prosociality in children. *Psychological* *Science*, *34*(4), 455–467. https://doi.org/10.1177/09567976221150616

艺术深深植根于人类的天性，并深刻地影响着人类的情感、思维和行为。每当我们欣赏精妙绝伦的艺术作品、历史悠久的人文名胜、鬼斧神工的自然景观时，都能获得与艺术交融汇通的精神上的满足，内心均会油然而生一种**敬畏之情**。  

**在个体面对相对于自身来说是巨大的并且超越了之前认识的自然和人造的物体，就可以引起敬畏的体验（Keltner & Haidt, 2003）。**  


![Image Name](https://cdn.kesci.com/upload/s69xd1u7hz.jpg?imageView2/0/w/640/h/640)  

在这种审美情趣的敬畏里，我们感受到到内心的平静和超然，而这与更健康的身体和更良好的社会功能息息相关(Stellar et al., 2017)。  

**研究发现，成年人的敬畏经历与更多的副交感神经系统（PNS）激活有关，PNS活动的增加（即RSA的增加）支持儿童和成年人平静的社会参与(Gordon et al., 2017)。**  

注：迷走神经控制呼吸性窦性心律不齐（RSA），即心率和呼吸的协同变化，这通常被认为是心脏健康的指标和情绪调节能力提升的预测因子 (Kok et al., 2013)

敬畏也可以**提升成年人的亲社会水平** (Keltner et al., 2014)。敬畏让人们感到自我的渺小、谦卑(Stellar et al., 2018)，并将注意力从自我转移到他人需求和社会世界上，从而导致人们做出更多的有益于他人的行为。  
**而在儿童身上我们也能看到类似的线索。**比如，5个月大的婴儿更容易对莫扎特的音乐微笑。  

而由于以往研究对积极情绪的忽视，并且假设敬畏这种复杂情绪更可能和个体后期发展阶段更同步，导致了**对童年期的敬畏研究不足**（Fredrickson., 2001)，值得注意的是，迄今为止还没有激发和测量儿童敬畏情绪的有效方法。

而在儿童的亲社会研究中发现，儿童帮助他人的内群体偏好在发展早期已经出现，比如，当与他人分享时，4-6岁的儿童对外群体接受者的慷慨分配少于内群体接受者；与来自另一所学校的孩子相比，3-8岁的孩子也更有可能与同学分享糖果。  
因此，本研究关注**艺术引起的敬畏对儿童对外群体的亲社会行为的影响及其独特的生理关联**。在两项预先登记的研究中(研究1:N = 159，研究2:N = 353)， 8 - 13岁的孩子观看了引起敬畏、喜悦或中性(对照)反应的电影片段。观看了这段令人敬畏的视频的孩子更有可能在亲社会任务中表现出更多的耐心(研究1)，并把他们的实验收入捐赠出去(研究1和2)用于帮助难民。他们还表现出增加的呼吸性窦性心律不齐，这是一个与社会参与有关的副交感神经系统激活的指标。我们讨论了通过重新想象儿童的环境来激发关键年龄阶段的敬畏，从而培养亲社会行为的意义。

## 研究一  


![Image Name](https://cdn.kesci.com/upload/s6docb2dk4.png?imageView2/0/w/320/h/320)   

**实验设计**：单因素被试间(诱发的情绪类型水平：积极情绪（快乐）、敬畏、中性情绪)  
**因变量指标**：计数数量、是否捐赠、  
**实验材料：**  
情绪诱发材料：电影《海之歌》(敬畏)、电影《幻想曲》(积极情绪)、如何粉刷墙壁的教学视频(中性)，每个片段时间为4分20秒  
情绪测量材料：使用六项表情符号量表报告了他们对视频的情绪反应，该量表包含对应于恐惧、愤怒、无聊、悲伤、喜悦和敬畏表情的表情符号;每个项目的反应选项范围从中性到非常强烈。  
**实验流程：**  
实验开始前，征得父母同意，得到同意后儿童学习如何播放视频、敬畏和喜悦之间的区别，并学会用5点李克特来报告情绪。完成练习课程后，将被试随机分到三种情绪状态下，在观看完视频后，完成情绪测量量表。随后对亲社会倾向进行测量。  
亲社会行为任务1：参与者被告知，当地一所大学组织了一次食品募捐活动，人们可以通过在清单上标记他们想要捐赠的食品数量来向难民捐赠食品。然而，必须对每个清单上的食品进行统计，以确保难民迅速获得食物。参与者被要求通过对大量捐赠清单进行分类来统计食物。该任务包括 30 份清单，其中包含七种不同的食物，每种食物最多可以捐赠 10 次。参与者被告知，完成这项任务是自愿的，他们可以随意停止计算食物。  
亲社会行为任务2：在计数任务结束后，参与者被告知实验结束，并获得了一张编号的抽奖券，这将允许他们参观当地的博物馆作为他们参与的奖励。然后，参与者被问及他们是否愿意将门票捐赠给难民家庭。

表情符号情绪测量量表  
![Image Name](https://cdn.kesci.com/upload/s6ml16ubvl.jpg)  


**研究假设：  
H1：观看诱发敬畏情绪的被试，比观看诱发中性和积极情绪视频的被试计数数量更多。  
H2：观看诱发敬畏情绪的被试，比观看诱发中性和积极情绪视频的被试更愿意捐赠门票。

数据筛选：删去未完成实验以及父母同意被撤回的被试数据

In [1]:
# 导入 pymc 模型包，和 arviz 等分析工具 
import pymc as pm
import arviz as az
import seaborn as sns
import scipy.stats as st
import numpy as np
import matplotlib.pyplot as plt
import xarray as xr
import pandas as pd

# 忽略不必要的警告
import warnings
warnings.filterwarnings("ignore")
df= pd.read_csv('/home/mw/input/dfghj6730/st111.csv')
df["emotioncondition_recorded"] = np.where(df['emotioncondition_recorded'] == 1, 0, np.where(df['emotioncondition_recorded'] == 3, 1, df['emotioncondition_recorded']))
df["emotioncondition_recorded"] = np.where(df['emotioncondition_recorded'] == 0, 1, np.where(df['emotioncondition_recorded'] == 1, 0, df['emotioncondition_recorded']))
df['emotioncondition_recorded'] = df['emotioncondition_recorded'].replace({0: 'Awe', 1: 'Joy', 2: 'Control'})
df["index"] = range(len(df))
df = df.set_index("index")

In [3]:
colors2 = ['red', 'blue', 'black']
sns.barplot(data=df,
            x = "emotioncondition_recorded", 
            y = "TotalItemCount",
            palette=colors2)
# 添加标题和x\y轴标签            
plt.xlabel('Emotion Condition')
plt.title('Item count in each emotion condition in Study 1')
plt.ylabel('Mean Item Count')

Text(0, 0.5, 'Mean Item Count')

<Figure size 640x480 with 1 Axes>

In [4]:
##study1中捐赠任务的条形图，计数
count_by = df.groupby(['emotioncondition_recorded', 'DonateTicket']).size().reset_index(name='Count')
print(count_by)

  emotioncondition_recorded  DonateTicket  Count
0                       Awe             1     37
1                       Awe             2     18
2                   Control             1     18
3                   Control             2     33
4                       Joy             1     22
5                       Joy             2     31


In [5]:
# 分类和组别
categories = ['Awe', 'Joy', 'Control']
groups = ['Yes', 'No']

# 数值
yes = [37, 22, 18]
no = [18, 31, 33]

# 设置柱状图的宽度和间距
bar_width = 0.35
bar_gap = 0.5

# 设置图形的大小
fig, ax = plt.subplots()

# 生成柱状图位置
index = np.arange(len(categories))

# 绘制 Yes 组别的柱状图
bar1 = ax.bar(index, yes, bar_width, label='Yes', color=['Red', 'Blue', 'Black'])

# 绘制 No 组别的柱状图
bar2 = ax.bar(index + bar_width, no, bar_width, label='No', color=['Red', 'Blue', 'Black'], alpha=0.7, hatch='//')

# 设置图形元素
ax.set_xlabel('Emotion Condition')
ax.set_ylabel('Ticket Donation (Count)')
ax.set_title(' Number of children who donated (Y = yes) and did not donate (N = no) in each emotion condition in Study 1')
ax.set_xticks(index + bar_width / 2)
ax.set_xticklabels(categories)
ax.legend()

# 显示数值标签
def autolabel(bars):
    for bar in bars:
        height = bar.get_height()
        ax.annotate('{}'.format(height),
                    xy=(bar.get_x() + bar.get_width() / 2, height),
                    xytext=(0, 3),  # 3 points vertical offset
                    textcoords="offset points",
                    ha='center', va='bottom')

autolabel(bar1)
autolabel(bar2)

# 显示图形
plt.show()

<Figure size 640x480 with 1 Axes>

# 亲社会行为——计数数量

## 自变量包含的类别：happy、awe

In [23]:
import warnings
warnings.filterwarnings("ignore")
df= pd.read_csv('/home/mw/input/3333121/31.csv')
#对自变量进行虚拟变量编码，将快乐编为0，敬畏编为1
df["EmotionCondition_recoded"] = np.where(df['EmotionCondition_recoded'] == 1, 0, np.where(df['EmotionCondition_recoded'] == 3, 1, df['EmotionCondition_recoded']))
df["index"] = range(len(df))
df = df.set_index("index")

## Poisson regression：  

1. 自变量：（emotion condition）快乐、敬畏  

2. 因变量：计数任务数量  

3. 数据关系：  
$$  
Y_i | \beta_0,\beta_1 \stackrel{ind}{\sim} Pois\left(\lambda_i\right) \;\; \text{ with } \;\; \lambda_i = e^{\beta_0 + \beta_1 X_{i1}}  
$$  

$$  
\begin{array}{lcrl}  
\text{priors:} & & \beta_{0}  &  \sim N\left(1, 1^2 \right)  \\  
               & & \beta_1  & \sim N\left(1, 1^2 \right)\\  
\end{array}  
$$  

### 泊松分布模型设定、先验选择

In [30]:
with pm.Model(coords = {"obs_id": df.index})  as poisson_model:

   beta_0 = pm.Normal("beta_0",mu=0.5,sigma=1)         #定义beta_0          
   beta_1=pm.Normal("beta_1",mu=0.5,sigma=1)           #定义beta_1
   #自变量:情绪唤醒条件
   emotion = pm.MutableData("emotion",df.EmotionCondition_recoded)
   ##定义mu，自变量与先验结合      
   lam = pm.Deterministic("lam", 
                           pm.math.exp(beta_0 + beta_1*emotion), 
                           dims="obs_id")  
   #定义泊松分布的似然
   likelihood = pm.Poisson("y_est", mu=lam, observed=df.TotalItemCountAccurate, dims="obs_id")

In [33]:
#使用 pm.sample_prior_predictive 函数从先验分布中生成预测样本
poisson_prior =pm.sample_prior_predictive(samples=50, 
                                          model=poisson_model,
                                          var_names=["lam"],
                                          random_seed=84735)

Sampling: [beta_0, beta_1]


### 先验检验

In [34]:
for i in range(50):
    plt.plot(poisson_prior.constant_data["emotion"],
             poisson_prior.prior["lam"].stack(sample=("chain", "draw"))[:, i], 
             c="grey",
             alpha=0.4)
# 设置x、y轴标题和总标题    
plt.xlabel("emotion", fontsize=12)
plt.ylabel("count", fontsize=12)
plt.suptitle("50 prior plausible models", fontsize=14)
sns.despine()
plt.show()


<Figure size 640x480 with 1 Axes>

### MCMC评估

In [35]:
with poisson_model:
    poisson_model_trace = pm.sample(
        draws=5000,
        tune=1000,
        chains=4,
        discard_tuned_samples=True,
        idata_kwargs={"log_likelihood": True},
        random_seed=84735
    )

Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta_0, beta_1]


Sampling 4 chains for 1_000 tune and 5_000 draw iterations (4_000 + 20_000 draws total) took 20 seconds.


In [36]:
az.plot_trace(poisson_model_trace, var_names=["beta_0","beta_1"],
              figsize=(15,10))
plt.show()

<Figure size 1500x1000 with 4 Axes>

In [37]:
az.summary(poisson_model_trace, var_names=["beta_0","beta_1"])

mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  ess_tail  \
beta_0  3.374  0.025   3.326    3.422        0.0      0.0    5874.0    8635.0   
beta_1  0.407  0.033   0.347    0.469        0.0      0.0    5704.0    8699.0   

        r_hat  
beta_0    1.0  
beta_1    1.0

### 后验预测评估

In [38]:
for i in range(100):
    sns.lineplot(x = poisson_model_trace.constant_data["emotion"],
            y = poisson_model_trace.posterior["lam"].stack(sample=("chain", "draw"))[:,i], 
            c="grey",
            alpha=0.4)
#设置x、y轴标题和总标题    
plt.xlabel("emotion",
           fontsize=12)
plt.ylabel("count",
           fontsize=12)
plt.suptitle("100 posterior plausible models",
           fontsize=14)
sns.despine()
plt.show()

<Figure size 640x480 with 1 Axes>

### 模型评估

In [86]:
#从后验分布中生成后验预测样本
with poisson_model:
    poisson_model_ppc = pm.sample_posterior_predictive(poisson_model_trace, random_seed=84735)

Sampling: [y_est]


In [88]:
az.plot_ppc(poisson_model_ppc, num_pp_samples=500)

<Axes: xlabel='y_est / y_est'>

<Figure size 640x480 with 1 Axes>

模型拟合非常不良

### 负二项回归模型设定、先验选择

##  Negative Binomial model：  

1. 自变量：（emotion condition）快乐、敬畏  

2. 因变量：计数任务数量  

3. 数据关系：  
$$  
\begin{equation}  
\begin{array}{lcrl}  
\text{data:} & \hspace{.025in} & Y_i|\beta_0,\beta_1,r & \stackrel{ind}{\sim} \text{NegBin}\left(\mu_i, r \right) \;\; \text{ with } \;\; \mu_i= e^{\beta_0 + \beta_1 X_{i1}}\\  
\text{priors:} & & \beta_{0}  & \sim N\left(1, 1,1^2\right) \\  
 & & \beta_1  & \sim N\left(1, 1^2\right) \\  
& & r & \sim \text{Exp}(1)\\  
\end{array}  
\end{equation}  
$$

In [52]:
with pm.Model(coords = {"obs_id": df.index}) as nb_model:

    beta_0 = pm.Normal("beta_0", mu=1,sigma=1)         #定义beta_0          
    beta_1=pm.Normal("beta_1", mu=1,sigma=1)           #定义beta_1
    alpha = pm.Exponential("alpha", 1)
    emotion = pm.MutableData("emotion",df.EmotionCondition_recoded)               
    lam = pm.Deterministic("lam", 
                           pm.math.exp(beta_0 + beta_1*emotion), 
                           dims="obs_id")    #定义mu，自变量与先验结合
    
    likelihood = pm.NegativeBinomial("y_est", mu=lam, alpha=alpha, observed=df.TotalItemCountAccurate)

In [53]:
pm.model_to_graphviz(nb_model)

### 先验检验

In [54]:
nb_prior = pm.sample_prior_predictive(samples=50, 
                                          model=nb_model,
                                          random_seed=84735)

Sampling: [alpha, beta_0, beta_1, y_est]


In [55]:
for i in range(50):
    plt.plot(nb_prior.constant_data["emotion"],
             nb_prior.prior["lam"].stack(sample=("chain", "draw"))[:, i], 
             c="grey",
             alpha=0.4)
# 设置x、y轴标题和总标题    
plt.xlabel("emotion", fontsize=13)
plt.ylabel("count", fontsize=13)
plt.suptitle("50 posterior plausible models", fontsize=13)
sns.despine()
plt.show()


<Figure size 640x480 with 1 Axes>

线条的预测方向比较一致，分布在0-40，并且从0-1呈现上升趋势，比较符合实际数据分布。

### MCMC评估

In [96]:
with nb_model:
    nb_model_trace = pm.sample(draws=5000,            # 使用mcmc方法进行采样，draws为采样次数
                      tune=1000,                    # tune为调整采样策略的次数，可以决定这些结果是否要被保留
                      chains=4,                     # 链数
                      discard_tuned_samples= True, 
                      idata_kwargs={"log_likelihood": True}, # tune的结果将在采样结束后被丢弃
                      random_seed=84735)

Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta_0, beta_1, alpha]


Sampling 4 chains for 1_000 tune and 5_000 draw iterations (4_000 + 20_000 draws total) took 29 seconds.


In [97]:
az.plot_trace(nb_model_trace, var_names=["beta_0","beta_1","alpha"],
              figsize=(15,15),
              compact=False)
plt.show()

<Figure size 1500x1500 with 6 Axes>

在后验检验中，每个参数的四条链比较重合

In [99]:
az.summary(nb_model_trace, var_names=["beta_0","beta_1"])

mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  ess_tail  \
beta_0  3.334  0.132   3.084    3.582      0.001    0.001   11762.0   12366.0   
beta_1  0.464  0.187   0.115    0.818      0.002    0.001   12331.0   12760.0   

        r_hat  
beta_0    1.0  
beta_1    1.0

r_hat=1，说明模型收敛良好

In [101]:
az.plot_posterior(nb_model_trace, var_names=["beta_0","beta_1"], transform = np.exp)
plt.show()

<Figure size 1472x552 with 2 Axes>

截距beta_0转换后的均值为28，置信区间不包含零。  
beta_1转换后的均值为1.6，置信区间不包含零。  
平均而言，当情绪从开心变为敬畏时，因变量增加约为1.8。

### 后验预测评估

In [102]:
for i in range(100):
    sns.lineplot(x = nb_model_trace.constant_data["emotion"],
            y = nb_model_trace.posterior["lam"].stack(sample=("chain", "draw"))[:,i], 
            c="grey",
            alpha=0.4)
#设置x、y轴标题和总标题    
plt.xlabel("emotion",
           fontsize=12)
plt.ylabel("count",
           fontsize=12)
plt.suptitle("100 posterior plausible models",
           fontsize=14)
sns.despine()
plt.show()

<Figure size 640x480 with 1 Axes>

后验预测结果趋势比较一致，主要集中在25-40,符合数据分布

### 模型评估

In [105]:
with nb_model:
    nb_model_ppc = pm.sample_posterior_predictive(nb_model_trace, random_seed=84735) 

az.plot_ppc(nb_model_ppc, num_pp_samples=500)

Sampling: [y_est]


<Axes: xlabel='y_est / y_est'>

<Figure size 640x480 with 1 Axes>

In [106]:
def count_outlier(i,hdi=0.94,ppc=nb_model_ppc):

  # 提取第i个X的值
  X = ppc.constant_data["emotion"][i].values
  # 提取第i个X值对应的后验预测值
  pre_ys = ppc.posterior_predictive["y_est"].stack(sample=("chain", "draw"))[i].values
  # 提取对应X值下的观测值Y
  true_ys = ppc.observed_data["y_est"][i].values
  
  # 计算对应后验预测的hdi的上下限
  lower,upper = az.hdi(pre_ys, hdi_prob=hdi)
  
  # 判断是否超过 hdi 边界
  outlier = (true_ys > upper) | (true_ys < lower)

  return outlier

In [107]:
hdi_verify = []

# 使用循环，判断所有X对应的Y是否超出后验预测范围，并计数
for i in range(len(df)):  
  hdi_verify.append(count_outlier(i, hdi=0.94))
total_number = sum(hdi_verify)
print(f"所有超过后验预测范围HDI的数量: {total_number}")

所有超过后验预测范围HDI的数量: 11


### 模型比较

In [108]:
comparison_list = {
    "nb_model":nb_model_trace,
    "ps_model":poisson_model_trace,
}
az.compare(comparison_list)

rank     elpd_loo      p_loo    elpd_diff        weight          se  \
nb_model     0  -499.508448   3.352833     0.000000  1.000000e+00   12.261822   
ps_model     1 -2232.806283  92.644228  1733.297835  9.725625e-10  291.583706   

                 dse  warning scale  
nb_model    0.000000    False   log  
ps_model  281.476041     True   log

负二项回归的elpd_loo的值更大，weight表示该模型在模型集合中相对于其他模型的相对支持程度。在这里，nb_model 的权重更大，表明它在模型集合中有很高的支持度。因此，在接下来的分析中，选择使用负二项回归模型。

接下来的分析中，除了对比对象的变化，其他都是与nb_model一致，所以接下来主要将我们的结果与假设对比，观察结果是否验证假设。

## 自变量包括的类别：awe、control

In [109]:
df= pd.read_csv('/home/mw/input/3333121/32.csv')
df["EmotionCondition_recoded"] = np.where(df['EmotionCondition_recoded'] == 2, 0, np.where(df['EmotionCondition_recoded'] == 3, 1, df['EmotionCondition_recoded']))
df["index"] = range(len(df))
df = df.set_index("index")

In [110]:
with pm.Model(coords = {"obs_id": df.index}) as nb_model1:

    beta_0 = pm.Normal("beta_0", mu=1, sigma=1)         #定义beta_0          
    beta_1=pm.Normal("beta_1", mu=1, sigma=1)
    alpha = pm.Exponential("alpha", 1)
    emotion = pm.MutableData("emotion",df.EmotionCondition_recoded)                     
                   
    lam1 = pm.Deterministic("lam1", 
                           pm.math.exp(beta_0 + beta_1*emotion), 
                           dims="obs_id")    
    
    likelihood = pm.NegativeBinomial("y_est", mu=lam1, alpha=alpha, observed=df.TotalItemCountAccurate)


![Image Name](https://cdn.kesci.com/upload/s6mfmr3322.png)  




![Image Name](https://cdn.kesci.com/upload/s6mfnz9grm.png)  




截距beta_0转换后的均值为20，置信区间不包含零。  
beta_1转换后的均值为2.2，置信区间不包含零。  
平均而言，当情绪从开心变为敬畏时，因变量增加约为2.2。

## 自变量包括的类别：happy、control

In [111]:
df= pd.read_csv('/home/mw/input/3333121/21.csv')
df["EmotionCondition_recoded"] = np.where(df['EmotionCondition_recoded'] == 2, 0, np.where(df['EmotionCondition_recoded'] == 1, 1, df['EmotionCondition_recoded']))
df["index"] = range(len(df))
df = df.set_index("index")

In [112]:
with pm.Model(coords = {"obs_id": df.index}) as nb_model2:

    beta_0 = pm.Normal("beta_0", mu=1,sigma=1)         #定义beta_0          
    beta_1=pm.Normal("beta_1", mu=1,sigma=1)       #定义beta_1
    alpha = pm.Exponential("alpha", 1)
    emotion = pm.MutableData("emotion",df.EmotionCondition_recoded)
    lam2 = pm.Deterministic("lam2", 
                           pm.math.exp(beta_0 + beta_1*emotion), 
                           dims="obs_id")    
    
    likelihood = pm.NegativeBinomial("y_est", mu=lam2, alpha=alpha, observed=df.TotalItemCountAccurate)

**beta_1的系数不显著，说明happy和control之间没有显著差别**  

![Image Name](https://cdn.kesci.com/upload/s6l3qpcln2.png)  


# 亲社会行为——捐赠

是否捐赠作为因变量是一个二分变量，所以应该使用逻辑回归

## 自变量包含类别：happy、awe     
## 因变量：是否捐赠

In [2]:
# 导入 pymc 模型包，和 arviz 等分析工具 
import pymc as pm
import arviz as az
import seaborn as sns
import scipy.stats as st
import numpy as np
import matplotlib.pyplot as plt
import xarray as xr
import pandas as pd

# 忽略不必要的警告
import warnings
warnings.filterwarnings("ignore")
df= pd.read_csv('/home/mw/input/3333121/31.csv')
df["EmotionCondition_recoded"] = np.where(df['EmotionCondition_recoded'] == 2, 0, np.where(df['EmotionCondition_recoded'] == 3, 1, df['EmotionCondition_recoded']))
df["DonateTicket"] = np.where(df['DonateTicket'] == 2,0,1)
df["index"] = range(len(df))
df = df.set_index("index")

## 逻辑回归-Bernoulli  

1. 自变量：（emotion condition）快乐、敬畏  

2. 因变量：捐赠行为  

3. 数据关系：  
$$  
\begin{array}{lcrl}  
\text{data:} & \hspace{.01in} & Y_i|\beta_0,\beta_1 & \stackrel{ind}{\sim} \text{Bern}(\pi_i) \;\; \text{ with } \;\; \pi_i = \frac{e^{\beta_0 + \beta_1 X_{i1}}}{1 + e^{\beta_0 + \beta_1 X_{i1}}} \\  
\text{priors:} & & \beta_{0}  &  \sim N\left(0.3, 0.4^2 \right)  \\  
               & & \beta_1  & \sim N\left(0.3, 0.4^2 \right)\\  
\end{array}  
$$

### 模型定义、先验选择

In [3]:
with pm.Model() as log_model1:
    # 此处对coords的定义方式进行了更改，因为后续我们需要进行对新数据的预测
    # 因此将维度定义成可更改的
    log_model1.add_coord('obs_id',df.index, mutable=True)

    y = pm.MutableData('y', df.DonateTicket, dims = 'obs_id')

    beta_0 = pm.Normal("beta_0",mu=0.5,sigma=0.5)          #定义beta_0          
    beta_1 = pm.Normal("beta_1",mu=0.5,sigma=0.5)          #定义beta_1
    emotion = pm.MutableData("emotion",df.EmotionCondition_recoded)                                            
    mu = pm.Deterministic("mu", 
                           beta_0 + beta_1*emotion, 
                           dims="obs_id")    
    #相当于进行了如下计算 (1 / (1 + exp(-mu))
    pi = pm.Deterministic("pi", pm.math.invlogit(mu), dims="obs_id")
    #似然
    likelihood = pm.Bernoulli("y_est",p=pi, observed=y,dims="obs_id")

In [4]:
pm.model_to_graphviz(log_model1)

### 先验检验

In [5]:
log1_prior = pm.sample_prior_predictive(samples=50, 
                                          model=log_model1,
                                          random_seed=84735)

Sampling: [beta_0, beta_1, y_est]


In [8]:
for i in range(log1_prior.prior.dims["draw"]):
    sns.lineplot(x = log1_prior.constant_data["emotion"],
                y = log1_prior.prior["pi"].stack(sample=("chain", "draw"))[:,i], c="grey" )
#设置x、y轴标题和总标题    
plt.xlabel("emotion",
           fontsize=12)
plt.ylabel("count",
           fontsize=12)
plt.suptitle("Relationships between emotion and count",
           fontsize=14)
sns.despine()
plt.show()

<Figure size 640x480 with 1 Axes>

先验趋势比较一致，主要分布在0.4-0.7，与实际情况比较相符

### MCMC评估

In [9]:
with log_model1:
    # MCMC 近似后验分布
    log_model1_trace = pm.sample(
                                draws=5000,                   # 使用mcmc方法进行采样，draws为采样次数
                                tune=1000,                    # tune为调整采样策略的次数，可以决定这些结果是否要被保留
                                chains=4,                     # 链数
                                discard_tuned_samples= True,  # tune的结果将在采样结束后被丢弃
                                random_seed=84735)
    log_model1_ppc = pm.sample_posterior_predictive(log_model1_trace)

Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta_0, beta_1]


In [10]:
az.plot_trace(log_model1_trace,
              var_names=["beta_0","beta_1"],
              figsize=(15,8),
              compact=False)
plt.show()

<Figure size 1500x800 with 4 Axes>

In [12]:
az.summary(log_model1_trace, var_names=["beta_0","beta_1"])

mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  ess_tail  \
beta_0 -0.054  0.293  -0.609    0.495      0.004    0.003    4998.0    7792.0   
beta_1 -0.049  0.295  -0.605    0.499      0.004    0.003    5160.0    7903.0   

        r_hat  
beta_0    1.0  
beta_1    1.0

In [179]:
az.plot_posterior(log_model1_trace, var_names=["beta_0","beta_1"], transform = np.exp)
plt.show()

<Figure size 1472x552 with 2 Axes>

In [180]:
for i in range(100):
    sns.lineplot(x = log_model1_trace.constant_data["emotion"],
            y = log_model1_trace.posterior["pi"].stack(sample=("chain", "draw"))[:,i], 
            c="grey",
            alpha=0.4)
#设置x、y轴标题和总标题    
plt.xlabel("emotion",
           fontsize=12)
plt.ylabel("probability of donate",
           fontsize=12)
plt.suptitle("100 posterior plausible models",
           fontsize=14)
sns.despine()
plt.show()

<Figure size 640x480 with 1 Axes>

### 模型评估

In [9]:
with log_model1:
    log_model1_ppc = pm.sample_posterior_predictive(log_model1_trace, random_seed=84735) 

az.plot_ppc(log_model1_ppc, num_pp_samples=500)

Sampling: [y_est]


<Axes: xlabel='y_est / y_est'>

<Figure size 640x480 with 1 Axes>

In [17]:
log_model1_ppc.posterior_predictive.y_est.stack(sample = ("chain", "draw"))

<xarray.DataArray 'y_est' (obs_id: 108, sample: 20000)>
array([[1, 0, 1, ..., 1, 1, 0],
       [0, 1, 1, ..., 0, 1, 1],
       [0, 1, 1, ..., 1, 0, 1],
       ...,
       [0, 1, 0, ..., 0, 0, 0],
       [1, 0, 0, ..., 0, 1, 0],
       [1, 0, 1, ..., 0, 1, 1]])
Coordinates:
  * obs_id   (obs_id) int64 0 1 2 3 4 5 6 7 ... 100 101 102 103 104 105 106 107
  * sample   (sample) object MultiIndex
  * chain    (sample) int64 0 0 0 0 0 0 0 0 0 0 0 0 ... 3 3 3 3 3 3 3 3 3 3 3 3
  * draw     (sample) int64 0 1 2 3 4 5 6 ... 4993 4994 4995 4996 4997 4998 4999

In [18]:
#stack(sample = ("chain", "draw")：将每一个X对应的4*5000个后验预测值合并到一个维度sample
#对于每一个X，需要计算其20000个值的平均值，因此将dim设置为sample
pred_pi = log_model1_ppc.posterior_predictive.y_est.stack(sample = ("chain", "draw")).mean(dim="sample")
# 转换为数据框
pred_pi = pred_pi.to_dataframe()

In [19]:
#将原数据中的X 和Y存入数据框
pred_pi["emotioncondition_recorded"] = log_model1_ppc.constant_data.emotion.values
pred_pi["DonateTicket"] = log_model1_ppc.observed_data.y_est.values

#根据分类标准（50-50）生成最终的分类结果
pred_pi["DonateTicket_pred"] = np.where(pred_pi["y_est"] >= 0.5, 1, 0)
pred_pi

y_est  emotioncondition_recorded  DonateTicket  DonateTicket_pred
obs_id                                                                     
0       0.49830                          0             0                  0
1       0.49095                          0             1                  0
2       0.49520                          0             0                  0
3       0.48475                          0             0                  0
4       0.48720                          0             0                  0
...         ...                        ...           ...                ...
103     0.63650                          1             0                  1
104     0.63700                          1             1                  1
105     0.63635                          1             1                  1
106     0.64070                          1             1                  1
107     0.63895                          1             1                  1

[108 rows x 4 columns]

In [20]:
confusion_matrix = pd.crosstab(pred_pi["DonateTicket"], pred_pi["DonateTicket_pred"], 
                              rownames=['Actual'], colnames=['Predicted'])
confusion_matrix

Predicted   0   1
Actual           
0          30  19
1          22  37

In [21]:
# 计算a b c d的数量
true_positive =  confusion_matrix.at[0, 1]
false_positive = confusion_matrix.at[1, 1]
true_negative = confusion_matrix.at[0,0]
false_negative = confusion_matrix.at[1,0]
# 代入公式
accuracy = (true_positive + true_negative) /(true_positive + false_positive + true_negative + false_negative)
sensitivity = (true_positive) /(true_positive + false_negative)
specificity = (true_negative) / (true_negative + false_positive)

print("准确性:", accuracy)
print("敏感性:", sensitivity)
print("特异性:", specificity)

准确性: 0.4537037037037037
敏感性: 0.4634146341463415
特异性: 0.44776119402985076


### 自变量包含的类别：awe、control

In [176]:
df= pd.read_csv('//home/mw/input/3333121/32.csv')
df["EmotionCondition_recoded"] = np.where(df['EmotionCondition_recoded'] == 2, 0, np.where(df['EmotionCondition_recoded'] == 3, 1, df['EmotionCondition_recoded']))
df["DonateTicket"] =  np.where(df['DonateTicket'] == 2, 0, 1)
df["index"] = range(len(df))
df = df.set_index("index")

In [181]:
with pm.Model() as log_model2:
    # 此处对coords的定义方式进行了更改，因为后续我们需要进行对新数据的预测
    # 因此将维度定义成可更改的
    log_model2.add_coord('obs_id',df.index, mutable=True)
    y = pm.MutableData('y', df.DonateTicket, dims = 'obs_id')
    beta_0 = pm.Normal("beta_0", mu=0.3, sigma=0.4)          #定义beta_0          
    beta_1 = pm.Normal("beta_1", mu=0.3, sigma=0.4)
    emotion = pm.MutableData("emotion",df.EmotionCondition_recoded)
                                        
    mu = pm.Deterministic("mu", 
                           beta_0 + beta_1*emotion, 
                           dims="obs_id")    
    #相当于进行了如下计算 (1 / (1 + exp(-mu))
    pi = pm.Deterministic("pi", pm.math.invlogit(mu), dims="obs_id")
    #似然
    likelihood = pm.Bernoulli("y_est",p=pi, observed=y,dims="obs_id")


![Image Name](https://cdn.kesci.com/upload/s7igjo996b.png?imageView2/0/w/960/h/960)  




![Image Name](https://cdn.kesci.com/upload/s6l588t4of.png)  


### 自变量包含的类别：hapyy、control

In [183]:
df= pd.read_csv('/home/mw/input/3333121/21.csv')
df["EmotionCondition_recoded"] = np.where(df['EmotionCondition_recoded'] == 2, 0, np.where(df['EmotionCondition_recoded'] == 1, 1, df['EmotionCondition_recoded']))
df["DonateTicket"] =  np.where(df['DonateTicket'] == 2, 0, 1)
df["index"] = range(len(df))
df = df.set_index("index")

In [184]:
with pm.Model() as log_model3:
    # 此处对coords的定义方式进行了更改，因为后续我们需要进行对新数据的预测
    # 因此将维度定义成可更改的
    log_model3.add_coord('obs_id',df.index, mutable=True)

    y = pm.MutableData('y', df.DonateTicket, dims = 'obs_id')
    beta_0 = pm.Normal("beta_0", mu=0.3, sigma=0.4)          #定义beta_0          
    beta_1 = pm.Normal("beta_1", mu=0.3, sigma=0.4)
    emotion = pm.MutableData("emotion",df.EmotionCondition_recoded)
                                            
    mu = pm.Deterministic("mu", 
                           beta_0 + beta_1*emotion, 
                           dims="obs_id")    
    
    #相当于进行了如下计算 (1 / (1 + exp(-mu))
    pi = pm.Deterministic("pi", pm.math.invlogit(mu), dims="obs_id")
    #似然
    likelihood = pm.Bernoulli("y_est",p=pi, observed=y,dims="obs_id")


![Image Name](https://cdn.kesci.com/upload/s6l5f8nr4r.png)  



研究1表明，敬畏促进了儿童的亲社会行为：观看令人敬畏的电影剪辑的儿童比观看令人快乐的电影剪辑或中立教学剪辑的儿童更有可能抽出时间并捐赠他们的实验收入来造福难民家庭。这些发现表明，敬畏促进了儿童的亲社会性。

# 研究二——RSA生理数据

study2与study1相比，增加了对生理指标的测量，其他过程与study1一致

### 描述性统计分析

In [187]:
# 读取数据
data_2 = pd.read_excel("/home/mw/project/研究二 神经数据.xlsx")

#验证被试总数 N = 353
num_participants = len(data_2['ppID'].unique())
print(num_participants)

#验证情感水平 = 3
emotion_levels = data_2['EmotionCondition'].unique()
print(emotion_levels)

#将自变量编码成虚拟变量 
# 自变量1   Emotion_2 = 0 Emotion_3 = 0
# 自变量2   Emotion_2 = 1 Emotion_3 = 0
# 自变量3   Emotion_2 = 0 Emotion_3 = 1

# 对EmotionCondition进行虚拟编码
emotion_dummies = pd.get_dummies(data_2['EmotionCondition'], prefix='Emotion')

# 将新的虚拟变量添加到原始数据框
df = pd.concat([data_2, emotion_dummies], axis=1)


# 重新编码 "ppID" 从1开始的新编码
df["ppID_new"] = pd.factorize(df["ppID"])[0] + 1

#加一列索引
df["obs_total"] =  range(1, len(df) + 1)

# 将ppID_new和Time列设置为索引
df.set_index([ "ppID_new", "obs_total","Time"], inplace=True, drop=False)

df

353
[1 3 2]


ppID  AnnualIncome  DispositionalHappy  \
ppID_new obs_total Time                                           
1        1         1        7           8.0                5.50   
         2         2        7           8.0                5.50   
         3         3        7           8.0                5.50   
         4         4        7           8.0                5.50   
         5         5        7           8.0                5.50   
...                       ...           ...                 ...   
353      3173      5     1191           8.0                3.33   
         3174      6     1191           8.0                3.33   
         3175      7     1191           8.0                3.33   
         3176      8     1191           8.0                3.33   
         3177      9     1191           8.0                3.33   

                         DispositionalAwe  ArtExposure  ParentArtInterest  \
ppID_new obs_total Time                                                     
1        1         1                 5.33          4.6               2.17   
         2         2                 5.33          4.6               2.17   
         3         3                 5.33          4.6               2.17   
         4         4                 5.33          4.6               2.17   
         5         5                 5.33          4.6               2.17   
...                                   ...          ...                ...   
353      3173      5                 2.83          2.8               3.50   
         3174      6                 2.83          2.8               3.50   
         3175      7                 2.83          2.8               3.50   
         3176      8                 2.83          2.8               3.50   
         3177      9                 2.83          2.8               3.50   

                         SubjectiveSES  baseline_mean_HR  baseline_mean_RSA  \
ppID_new obs_total Time                                                       
1        1         1              5.83             69.73               8.23   
         2         2              5.83             69.73               8.23   
         3         3              5.83             69.73               8.23   
         4         4              5.83             69.73               8.23   
         5         5              5.83             69.73               8.23   
...                                ...               ...                ...   
353      3173      5              5.17            105.38               7.28   
         3174      6              5.17            105.38               7.28   
         3175      7              5.17            105.38               7.28   
         3176      8              5.17            105.38               7.28   
         3177      9              5.17            105.38               7.28   

                         baseline_mean_SCL  ...    HR    RSA    SCL  SCL_SQ  \
ppID_new obs_total Time                     ...                               
1        1         1                  9.87  ...  64.8  8.328  13.70    3.70   
         2         2                  9.87  ...  63.1  8.355  14.43    3.80   
         3         3                  9.87  ...  67.7  8.421  14.17    3.76   
         4         4                  9.87  ...  63.3  8.208  13.50    3.67   
         5         5                  9.87  ...  64.8  8.356  13.00    3.61   
...                                    ...  ...   ...    ...    ...     ...   
353      3173      5                 17.93  ...  88.1  7.296  19.93    4.46   
         3174      6                 17.93  ...  88.7  7.793  19.83    4.45   
         3175      7                 17.93  ...  79.3  6.672  20.02    4.47   
         3176      8                 17.93  ...  83.9  7.674  19.80    4.45   
         3177      9                 17.93  ...  81.5  5.383  20.00    4.47   

                         filter_$  Emotion_1  Emotion_2  Emotion_3  ppID_new  \
ppID_new obs_to

In [188]:
#RSA数据可视化 核密度图x：emotional condition y：RSA

# 提取分类变量和连续变量的列
category_variable_column = df['EmotionCondition']
continuous_variable_column = df['RSA']

# 使用Seaborn绘制核密度估计图
plt.figure(figsize=(8, 6))
sns.kdeplot(data=df, x=continuous_variable_column, hue=category_variable_column, fill=True)
plt.title('不同情绪状态下的RSA核分布图')
plt.xlabel('RSA')
plt.ylabel('Density')
# 添加图例 1 = awe， 2 = joy， 3 = control
plt.legend(title='EmotionCondition', labels=['Awe', 'Joy', 'Control'])
plt.show()

<Figure size 800x600 with 1 Axes>

In [11]:
#RSA数据可视化_折线图 

# 指定颜色
plt.figure(figsize=(12, 6)) 

# 指定颜色 1：Awe，2：joy，3：control
colors = {1: 'red', 2: 'blue', 3: 'black'}

sns.lineplot(x="Time", y="RSA", hue="EmotionCondition", data=df, ci=None, palette=colors, hue_order=[1, 2, 3])

plt.title('RSA over Time by EmotionCondition')
plt.xlabel('Time')
plt.ylabel('RSA')

# 添加图例
plt.legend(title='EmotionCondition', labels=['Awe', 'Joy', 'Control'])

plt.show()

<Figure size 1200x600 with 1 Axes>

In [42]:
# 读取数据
df_subset = pd.read_excel("/home/mw/project/研究二 神经（9个被试）.xlsx")

#将自变量编码成虚拟变量 
df_subset['Emotion_1'] = df_subset['EmotionCondition'].replace({1: 1, 3: 0})

# 重新编码 "ppID" 从1开始的新编码
df_subset["ppID_new"] = pd.factorize(df_subset["ppID"])[0] + 1

#加一列索引
df_subset["obs_total"] =  range(1, len(df_subset) + 1)

# 将ppID_new和Time列设置为索引
df_subset.set_index([ "ppID_new", "obs_total","Time"], inplace=True, drop=False)


### 贝叶斯层次模型  

自变量：1情绪状态 awe情绪条件  control情绪条件  2时间  
因变量：RSA水平  

![Image Name](https://cdn.kesci.com/upload/s6l9l35f2o.png?imageView2/0/w/960/h/960)  


创建三个模型，比较其拟合程度

### 模型1：RSA的固定截距模型  
$$  
[ \text{RSA}_{\text{fixedintercept}} = \beta_0 + \beta_1 \times \text{Time} + \beta_2 \times \text{EmotionCondition} + \beta_3 \times \text{baseline_mean_RSA} + \beta_4 \times \text{practice_mean_RSA} + \beta_5 \times \text{Time} \times \text{EmotionCondition} + \epsilon ]  
$$

In [44]:
# 创建固定截距RSA模型
with pm.Model() as RSA_fixedintercept_model:

    # 模型设定
    beta_0 = pm.Normal("beta_0", mu=0, sigma=1)  
    beta_time = pm.Normal("beta_time", mu=0, sigma=1)
    beta_emotion_2 = pm.Normal("beta_emotion_2", mu=0, sigma=1)
    beta_emotion_4 = pm.Normal("beta_emotion_4", mu=0, sigma=1)
    beta_baseline_RSA = pm.Normal("beta_baseline_RSA", mu=0, sigma=1)
    beta_practice_RSA = pm.Normal("beta_practice_RSA", mu=0, sigma=1)
    

    # 线性模型
    mu = pm.Deterministic("mu",
        beta_0 +
        beta_time * df_subset["Time"]+
        beta_emotion_2 * df_subset["Emotion_1"] +
        beta_emotion_4 * df_subset["Emotion_1"] * df_subset["Time"]+
        beta_baseline_RSA * df_subset["baseline_mean_RSA"] +
        beta_practice_RSA * df_subset["practice_mean_RSA"]
    )
    
    # 定义观测模型
    sigma = pm.HalfCauchy("sigma", beta=1)
    likelihood = pm.Normal("y_RSA", mu=mu, sigma=sigma, observed=df_subset["RSA"])



In [ ]:

    RSA_fixedintercept_trace = pm.sample(draws=1000,
                                    tune=300,
                                    chains=4,
                                    discard_tuned_samples=True,
                                    random_seed=84735,
                                    target_accept=0.99)

### 模型2：RSA的随机截距模型（被试随机截距）  
$$  
[ \text{RSA}_{\text{randomintercept}} = \beta_0 + \beta_1 \times \text{Time} + \beta_2 \times \text{EmotionCondition} + \beta_3 \times \text{baseline_mean_RSA} + \beta_4 \times \text{practice_mean_RSA} + \beta_5 \times \text{Time} \times \text{EmotionCondition} + \text{random_intercept} + \epsilon ]  
$$

In [198]:
def run_RSA_randomintercept_model(non_centered=False):

    # 定义数据坐标，包括被试和观测索引
    coords = {"ppID": df_subset['ppID_new'].values,  # 使用 .values 属性将 Pandas Series 转换为 NumPy 数组
              "obs_total": df_subset['obs_total'].values}
      
    with pm.Model(coords=coords) as RSA_randomintercept_model:
        # 定义全局参数
        beta_0 = pm.Normal("beta_0", mu=5, sigma=1)
        beta_0_sigma = pm.Exponential("beta_0_sigma", 1)
        beta_1 = pm.Normal("beta_1", mu=0, sigma=1)
        sigma_y = pm.Exponential("sigma_y", 1) 
        beta_3 = pm.Normal("beta_3", mu=0, sigma=1)
        beta_5 = pm.Normal("beta_5", mu=0, sigma=1)
        beta_6 = pm.Normal("beta_6", mu=0, sigma=1)
        ppID = df_subset['ppID_new'].values

        # 被试随机效应
        if non_centered:
            beta_0_offset = pm.Normal("beta_0_offset", 0, sigma=1, dims="ppID")
            beta_0j = pm.Deterministic("beta_0j", beta_0 + beta_0_offset * beta_0_sigma, dims="ppID")
        else:
            beta_0j = pm.Normal("beta_0j", mu=beta_0, sigma=beta_0_sigma, dims="ppID")

        # 线性关系
        mu = pm.Deterministic("mu", 
                              beta_0j[ppID] +
                              beta_1 * df_subset['Emotion_1'] +
                              beta_3 * df_subset['Time'] * df_subset['Emotion_1'] +
                              beta_5 * df_subset['baseline_mean_RSA'] +
                              beta_6 * df_subset['practice_mean_RSA'],            
                              dims="obs_total")

        # 定义 likelihood
        likelihood = pm.Normal("y_RSA", mu=mu, sigma=sigma_y, observed=df_subset.RSA, dims="obs_total")

        RSA_randomintercept_trace = pm.sample(draws=1000,
                                             tune=300,
                                             chains=4,
                                             discard_tuned_samples=True,
                                             random_seed=84735,
                                             target_accept=0.99)
    
    return RSA_randomintercept_model, RSA_randomintercept_trace



In [199]:
RSA_randomintercept_model, RSA_randomintercept_trace = run_RSA_randomintercept_model(non_centered = True)

Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta_0, beta_0_sigma, beta_1, sigma_y, beta_3, beta_5, beta_6, beta_0_offset]


Sampling 4 chains for 300 tune and 1_000 draw iterations (1_200 + 4_000 draws total) took 558 seconds.


### 模型3：一阶自回归模型  
$$  
 [ \text{RSA}_{\text{randomAR1}} = \beta_0 + \beta_1 \times \text{Time} + \beta_2 \times \text{EmotionCondition} + \beta_3 \times \text{baseline_mean_RSA} + \beta_4 \times \text{practice_mean_RSA} + \beta_5 \times \text{Time} \times \text{EmotionCondition} + \text{random_effects} + \text{correlation_structure} + \epsilon ]  
$$  

level1:被试水平（不同被试的随机截距）  

level2:时间水平（同一被试的不同时间点1-9的时间点，由于每一时间点的数据都和上一个数据相关，因此使用一阶自回归模型，用来表示当前时刻的观测值与上一个时刻的观测值之间存在线性关系）  



In [200]:
def run_RSA_randomAR1_model(non_centered=False):
    # 定义数据坐标，包括被试和观测索引

    coords = {"ppID": df_subset['ppID_new'].values,  
              "obs_total": df_subset['obs_total'].values}

    with pm.Model(coords=coords) as RSA_randomAR1_model:
        # 定义全局参数
        beta_0 = pm.Normal("beta_0", mu=5, sigma=1)
        beta_0_sigma = pm.Exponential("beta_0_sigma", 1)
        beta_time = pm.Normal("beta_time", mu=0, sigma=5)
        beta_emotion_2 = pm.Normal("beta_emotion_2", mu=0, sigma=5)
        beta_time_emotion_2 = pm.Normal("beta_time_emotion_2", mu=0, sigma=5)
        baseline_mean_RSA = pm.Normal("baseline_mean_RSA", mu=0, sigma=5)
        practice_mean_RSA = pm.Normal("practice_mean_RSA", mu=0, sigma=5)
        ppID = df_subset['ppID_new'].values
        
        # 被试随机效应
        if non_centered:
            beta_0_offset = pm.Normal("beta_0_offset", 0, sigma=1, dims="ppID")
            beta_0j = pm.Deterministic("beta_0j", beta_0 + beta_0_offset * beta_0_sigma, dims="ppID")
        else:
            beta_0j = pm.Normal("beta_0j", mu=beta_0, sigma=beta_0_sigma, dims="ppID")

        # 定义时间效应的随机效应（一阶自回归）
        rho = pm.Normal("rho", mu=0, sigma=1, dims="ppID")
        sigma_time = pm.Exponential("sigma_time", 1, dims="ppID")
        time_effect = rho[ppID] * df_subset["Time"].diff().fillna(0)  # 时间差分


        # 线性关系
        mu = pm.Deterministic(
            "mu", 
            beta_0j[ppID] + 
            beta_time * df_subset["Time"] +
            beta_emotion_2 * df_subset["Emotion_1"] + 
            beta_time_emotion_2 * df_subset["Time"] * df_subset["Emotion_1"] +
            baseline_mean_RSA * df_subset["baseline_mean_RSA"] +
            practice_mean_RSA * df_subset["practice_mean_RSA"] +
            time_effect,
            dims="obs_total"
        )

        # 定义 likelihood
        likelihood = pm.Normal("y_RSA", mu=mu, sigma=beta_0_sigma, observed=df_subset["RSA"], dims="obs_total")


        # 进行采样
        RSA_randomAR1_trace = pm.sample(draws=1000, 
                                        tune=300, 
                                        chains=4, 
                                        discard_tuned_samples=True, 
                                        random_seed=84735, 
                                        target_accept=0.99)

    return RSA_randomAR1_model, RSA_randomAR1_trace


In [201]:
RSA_randomAR1_model, RSA_randomAR1_trace = run_RSA_randomAR1_model(non_centered = True)

Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta_0, beta_0_sigma, beta_time, beta_emotion_2, beta_time_emotion_2, baseline_mean_RSA, practice_mean_RSA, beta_0_offset, rho, sigma_time]


Sampling 4 chains for 300 tune and 1_000 draw iterations (1_200 + 4_000 draws total) took 1046 seconds.


### 模型比较

### 后验预测检查

In [31]:
with RSA_fixedintercept_model:
    RSA_fixedintercept_model_ppc = pm.sample_posterior_predictive(RSA_fixedintercept_trace, random_seed=84735)
with RSA_randomintercept_model:
    RSA_randomintercept_model_ppc = pm.sample_posterior_predictive(RSA_randomintercept_trace, random_seed=84735)
with RSA_randomAR1_model:
    RSA_randomAR1_model_ppc = pm.sample_posterior_predictive(RSA_randomAR1_trace, random_seed=84735)

Sampling: [y_RSA]


Sampling: [y_RSA]


Sampling: [y_RSA]


In [32]:

# 定义计算 MAE 函数
from statistics import median
def MAE(model_ppc):
    # 计算每个X取值下对应的后验预测模型的均值
    pre_y = model_ppc.posterior_predictive["y_RSA"].stack(sample=("chain", "draw"))
    pre_y_mean = pre_y.mean(axis=1).values

    # 提取观测值Y，提取对应Y值下的后验预测模型的均值
    MAE = pd.DataFrame({
        "RSA_ppc_mean": pre_y_mean,
        "RSA_original": df_subset.RSA
    })

    # 计算预测误差
    MAE["pre_error"] = abs(MAE["RSA_original"] -\
                            MAE["RSA_ppc_mean"])

    # 最后，计算预测误差的中位数
    MAE = median(MAE.pre_error)
    return MAE

In [33]:
# 定义计算超出HDI的数量
def counter_outlier(model_ppc, hdi_prob=0.95):
    # 将az.summary生成的结果存到hdi_multi这个变量中，该变量为数据框
    hdi = az.summary(model_ppc, kind="stats", hdi_prob=hdi_prob)
    lower = hdi.iloc[:,2].values
    upper = hdi.iloc[:,3].values

    # 将原数据中的RSA数据合并，便于后续进行判断
    y_obs = model_ppc.observed_data["y_RSA"].values

    # 判断原数据中的RSA分数是否在后验预测的95%可信区间内，并计数
    hdi["verify"] = (y_obs <= lower) | (y_obs >= upper)
    hdi["y_obs"] = y_obs
    hdi_num = sum(hdi["verify"])

    return hdi_num

In [34]:
#比较MAE
RSA_fixedintercept_model_MAE = MAE(RSA_fixedintercept_model_ppc)
print(f"固定截距模型MAE: {RSA_fixedintercept_model_MAE:.2f}")

RSA_randomintercept_model_MAE = MAE(RSA_randomintercept_model_ppc)
print(f"随机截距模型MAE: {RSA_randomintercept_model_MAE:.2f}")

RSA_randomAR1_model_MAE = MAE(RSA_randomAR1_model_ppc)
print(f"一阶自回归模型MAE: {RSA_randomAR1_model_MAE:.2f}")


固定截距模型MAE: 0.28
随机截距模型MAE: 0.21
一阶自回归模型MAE: 0.17


### MAE结果解读  
MAE（平均绝对误差）数值越小表示模型的预测误差越小，模型的性能越好。  

固定截距模型的 MAE 是 0.28，随机截距模型的 MAE 是 0.21，一阶自回归模型的 MAE 是 0.17。根据这些数值来看，一阶自回归模型的 MAE 最小，因此在这个比较中，一阶自回归模型的性能最好。

In [35]:
#比较超出HDI的数量
RSA_fixedintercept_model_outliers = counter_outlier(RSA_fixedintercept_model_ppc)
print(f"固定截距模型超出95%hdi: {RSA_fixedintercept_model_outliers:.2f}")

RSA_randomintercept_model_outliers = counter_outlier(RSA_randomintercept_model_ppc)
print(f"随机截距模型超出95%hdi: {RSA_randomintercept_model_outliers:.2f}")

RSA_randomAR1_model_outliers = counter_outlier(RSA_randomAR1_model_ppc)
print(f"一阶自回归模型超出95%hdi: {RSA_randomAR1_model_outliers:.2f}")

固定截距模型超出95%hdi: 1.00
随机截距模型超出95%hdi: 3.00
一阶自回归模型超出95%hdi: 1.00


### HDI结果解读  
95% HDI（高密度区间，High Density Interval）表示对参数的不确定性范围。当HDIs超出1.0时，意味着这些模型的某些参数在95%的置信水平下超出了较宽的不确定性范围。  

固定截距模型的某些参数超出了95% HDI，随机截距模型的某些参数超出了3.00个单位，而一阶自回归模型的某些参数超出了1.00个单位。  

这意味着在这些模型中，一些参数可能存在较大的不确定性，因为它们的后验分布（posterior distribution）在95%的情况下覆盖了较宽的范围。  





In [47]:
for trace in [RSA_fixedintercept_trace, RSA_randomintercept_trace, RSA_randomAR1_trace]:
    if hasattr(trace, 'log_likelihood'):
        del trace.log_likelihood  # 移除已存在的 log_likelihood 组

# 计算模型对数似然值以便进一步计算 ELPD. 
with RSA_fixedintercept_model:
    pm.compute_log_likelihood(RSA_fixedintercept_trace)

with RSA_randomintercept_model:
   pm.compute_log_likelihood(RSA_randomintercept_trace)

with RSA_randomAR1_model:
   pm.compute_log_likelihood(RSA_randomAR1_trace)



In [49]:
comparison_list = {
    "RSA_fixedintercept_model":RSA_fixedintercept_trace,
    "RSA_randomintercept_model":RSA_randomintercept_trace,
    "RSA_randomAR1_model":RSA_randomAR1_trace,
}
az.compare(comparison_list)

rank   elpd_loo      p_loo  elpd_diff    weight  \
RSA_fixedintercept_model      0 -36.178357   5.947685   0.000000  0.567596   
RSA_randomintercept_model     1 -37.389660   7.181564   1.211304  0.432404   
RSA_randomAR1_model           2 -41.065668  12.062097   4.887312  0.000000   

                                 se       dse  warning scale  
RSA_fixedintercept_model   4.914405  0.000000    False   log  
RSA_randomintercept_model  6.734843  3.560521    False   log  
RSA_randomAR1_model        5.232833  2.592997     True   log

### 模型比较结果  
RSA_fixedintercept_model的elpd_loo值最大，为-36.178357，拟合结果最好  
RSA_randomAR1_model的elpd_loo值最小，为-41.065668，拟合最不好

In [195]:
az.summary(RSA_fixedintercept_trace, var_names=['beta_0', 'beta_time', 'beta_emotion_2', 'beta_emotion_4', 'beta_baseline_RSA', 'beta_practice_RSA'])

mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  \
beta_0             0.891  0.507  -0.043    1.895      0.011    0.008   
beta_time          0.062  0.033   0.002    0.126      0.001    0.001   
beta_emotion_2     0.461  0.268  -0.043    0.961      0.007    0.005   
beta_emotion_4    -0.084  0.046  -0.173    0.002      0.001    0.001   
beta_baseline_RSA  1.059  0.268   0.557    1.583      0.007    0.005   
beta_practice_RSA -0.207  0.291  -0.748    0.363      0.008    0.005   

                   ess_bulk  ess_tail  r_hat  
beta_0               2200.0    2284.0    1.0  
beta_time            1741.0    2021.0    1.0  
beta_emotion_2       1615.0    2203.0    1.0  
beta_emotion_4       1642.0    1900.0    1.0  
beta_baseline_RSA    1522.0    1662.0    1.0  
beta_practice_RSA    1481.0    1832.0    1.0

In [212]:
az.summary(RSA_randomintercept_trace,var_names=['beta_0', 'beta_1', 'beta_3', 'beta_5', 'beta_6'])

mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  ess_tail  \
beta_0  3.772  1.081   1.886    5.780      0.030    0.021    1303.0    2287.0   
beta_1  0.266  0.474  -0.630    1.200      0.012    0.009    1723.0    1552.0   
beta_3 -0.025  0.033  -0.086    0.039      0.000    0.000    5796.0    2861.0   
beta_5  0.878  0.590  -0.282    1.935      0.016    0.011    1348.0    2054.0   
beta_6 -0.393  0.557  -1.413    0.731      0.014    0.010    1663.0    1980.0   

        r_hat  
beta_0    1.0  
beta_1    1.0  
beta_3    1.0  
beta_5    1.0  
beta_6    1.0

In [218]:
az.summary(RSA_randomAR1_trace,var_names=['beta_0', 'beta_time', 'beta_emotion_2', 'beta_time_emotion_2', 'baseline_mean_RSA', 'practice_mean_RSA'])

mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  \
beta_0               4.094  0.905   2.455    5.854      0.015    0.011   
beta_time            0.087  0.040   0.011    0.160      0.001    0.001   
beta_emotion_2       0.434  0.582  -0.694    1.484      0.017    0.012   
beta_time_emotion_2 -0.106  0.057  -0.207    0.005      0.001    0.001   
baseline_mean_RSA    1.776  0.989  -0.173    3.537      0.027    0.019   
practice_mean_RSA   -1.369  0.986  -3.228    0.496      0.026    0.019   

                     ess_bulk  ess_tail  r_hat  
beta_0                 3785.0    2946.0   1.00  
beta_time              2618.0    2541.0   1.00  
beta_emotion_2         1127.0    1539.0   1.00  
beta_time_emotion_2    2321.0    2951.0   1.00  
baseline_mean_RSA      1378.0    1873.0   1.01  
practice_mean_RSA      1413.0    1816.0   1.01

Fredrickson, B. L. (2001). The role of positive emotions in positive psychology: The broaden-and-build theory of positive emotions. American Psychologist, 56(3), 218–226. https://doi.org/10.1037/0003-066x.56.3.218  
Gordon, A. M., Stellar, J. E., Anderson, C. L., McNeil, G. D., Loew, D., & Keltner, D. (2017). The dark side of the sublime: Distinguishing a threat-based variant of awe. Journal of Personality and Social Psychology, 113(2), 310–328. https://doi.org/10.1037/pspp0000120  
Keltner, D., & Haidt, J. (2003). Approaching awe, a moral, spiritual, and aesthetic emotion. Cognition & Emotion, 17(2), 297–314. https://doi.org/10.1080/02699930302297  
Keltner, D., Kogan, A., Piff, P. K., & Saturn, S. R. (2014). The Sociocultural Appraisals, Values, and Emotions (SAVE) Framework of Prosociality: Core Processes from Gene to Meme. Annual Review of Psychology, 65(1), 425–460. https://doi.org/10.1146/annurev-psych-010213-115054  
Kok, B. E., Coffey, K. A., Cohn, M., Catalino, L. I., Vacharkulksemsuk, T., Algoe, S. B., Brantley, M. M., & Fredrickson, B. L. (2013). How positive emotions build physical health. Psychological Science, 24(7), 1123–1132. https://doi.org/10.1177/0956797612470827  
Stellar, J. E., Gordon, A. M., Anderson, C. L., Piff, P. K., McNeil, G. D., & Keltner, D. (2018). Awe and humility. Journal of Personality and Social Psychology, 114(2), 258–269. https://doi.org/10.1037/pspi0000109  
Stellar, J. E., Gordon, A. M., Piff, P. K., Cordaro, D. T., Anderson, C. L., Bai, Y., Maruskin, L. A., & Keltner, D. (2017). Self-Transcendent emotions and their social functions: compassion, gratitude, and awe bind us to others through prosociality. Emotion Review, 9(3), 200–207. https://doi.org/10.1177/1754073916684557  
